nixos/modules/services/networking/fastnetmon-advanced.nix

```diff
@@ -0,0 +1,193 @@
+
+    mkdir -p /etc/fastnetmon/keychain
+    cat << EOF > /etc/fastnetmon/fastnetmon.conf
+    ${builtins.toJSON {
+      mongodb_username = "";
+    }}
+    EOF
+    ${pkgs.fastnetmon-advanced}/bin/fcli create_configuration
+    ${pkgs.fastnetmon-advanced}/bin/fcli set bgp default
+    ${pkgs.fastnetmon-advanced}/bin/fcli export_configuration backup.tar
+    tar -C $out --no-same-owner -xvf backup.tar
+  '');
+
+  # merge the user configs into the default configs
+  config_tar = pkgs.runCommand "fastnetmon-config.tar" {
+    nativeBuildInputs = with pkgs; [ jq ];
+  } ''
+    jq -s add ${default_configs}/main.json ${pkgs.writeText (baseNameOf "main-add.json") (builtins.toJSON cfg.settings)} > main.json
+    mkdir hostgroup
+    ${lib.concatImapStringsSep "\n" (pos: hostgroup: ''
+      jq -s add ${default_configs}/hostgroup/0.json ${pkgs.writeText (baseNameOf "hostgroup-${toString (pos - 1)}-add.json") (builtins.toJSON hostgroup)} > hostgroup/${toString (pos - 1)}.json
+    '') hostgroups}
+    mkdir bgp
+    ${lib.concatImapStringsSep "\n" (pos: bgp: ''
+      jq -s add ${default_configs}/bgp/0.json ${pkgs.writeText (baseNameOf "bgp-${toString (pos - 1)}-add.json") (builtins.toJSON bgp)} > bgp/${toString (pos - 1)}.json
+    '') bgpPeers}
+    tar -cf $out main.json ${lib.concatImapStringsSep " " (pos: _: "hostgroup/${toString (pos - 1)}.json") hostgroups} ${lib.concatImapStringsSep " " (pos: _: "bgp/${toString (pos - 1)}.json") bgpPeers}
+  '';
+
+  hostgroups = lib.mapAttrsToList (name: hostgroup: { inherit name; } // hostgroup) cfg.hostgroups;
+  bgpPeers = lib.mapAttrsToList (name: bgpPeer: { inherit name; } // bgpPeer) cfg.bgpPeers;
+
+in {
+  options.services.fastnetmon-advanced = with lib; {
+    enable = mkEnableOption "the fastnetmon-advanced DDoS Protection daemon";
+
+    settings = mkOption {
+      description = "Extra configuration options to declaratively load into FastNetMon Advanced";
+      type = settingsFormat.type;
+      default = {};
+    };
+    hostgroups = mkOption {
+      description = "Hostgroups to declaratively load into FastNetMon Advanced";
+      type = types.attrsOf settingsFormat.type;
+      default = {};
+    };
+    bgpPeers = mkOption {
+      description = "BGP Peers to declaratively load into FastNetMon Advanced";
+      type = types.attrsOf settingsFormat.type;
+      default = {};
+    };
+
+    enableAdvancedTrafficPersistence = mkOption {
+      description = "Store historical flow data in clickhouse";
+      type = types.bool;
+      default = false;
+    };
+  };
+
+  config = lib.mkMerge [ (lib.mkIf cfg.enable {
+    environment.systemPackages = with pkgs; [
+      fastnetmon-advanced
+    ];
+
+    environment.etc."fastnetmon/license.lic".source = "/var/lib/fastnetmon/license.lic";
+    environment.etc."fastnetmon/gobgpd.conf".source = "/run/fastnetmon/gobgpd.conf";
```

Comment:
You understand correctly. We want these stateful files to be not in `/etc`.

`license.lic` is the cached fastnetmon advanced license. In fastnetmon advanced, a license is a assigned on the licensing server to an IP address. When fastnetmon advanced starts up, it attempts to obtain the license from the license server, and saves it to `/etc/fastnetmon/license.lic` which is valid for $TIME (14 days or so) until it has to obtain a renewed license. Think similar to how you would obtain TLS certificates over ACME and renew them periodically.

Reply:
Oh okay then that's fine, I thought it might be like a license agreement which you accept by moving the file or something :see_no_evil: which then we would have needed to add a option for accept terms like in the ACME module. While on the topic, does fastnetmon-advanced have a eula the user automatically accepts by using this NixOS module?

Reply:
I guess there is this: https://fastnetmon.com/terms-conditions/

Reply:
But the package is already as licenses.unfree, so users are aware they are selling their souls

Reply:
I guess the license already applys when downloading the package so going from installing the package to using the module probably doesn't really change anything. So I'm fine leaving it as is.